networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/rpc/RpcHandler.java

```diff
@@ -0,0 +1,152 @@
+import org.jetbrains.annotations.NotNull;
+import tech.pegasys.artemis.networking.p2p.libp2p.LibP2PNodeId;
+import tech.pegasys.artemis.networking.p2p.libp2p.rpc.RpcHandler.Controller;
+import tech.pegasys.artemis.networking.p2p.peer.NodeId;
+import tech.pegasys.artemis.networking.p2p.rpc.RpcMethod;
+import tech.pegasys.artemis.networking.p2p.rpc.RpcRequestHandler;
+import tech.pegasys.artemis.networking.p2p.rpc.RpcStream;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+public class RpcHandler implements ProtocolBinding<Controller> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final RpcMethod rpcMethod;
+
+  public RpcHandler(RpcMethod rpcMethod) {
+    this.rpcMethod = rpcMethod;
+  }
+
+  @SuppressWarnings("unchecked")
+  public SafeFuture<RpcStream> sendRequest(
+      Connection connection, Bytes initialPayload, RpcRequestHandler handler) {
+    return SafeFuture.of(
+            connection
+                .getMuxerSession()
+                .createStream(
+                    Multistream.create(this.toInitiator(rpcMethod.getId())).toStreamHandler())
+                .getControler())
+        .thenCompose(
+            ctr -> {
+              ctr.setRequestHandler(handler);
+              return ctr.getRpcStream()
+                  .writeBytes(initialPayload)
+                  .thenApply(f -> ctr.getRpcStream());
+            });
+  }
+
+  @NotNull
+  @Override
+  public String getAnnounce() {
+    return rpcMethod.getId();
+  }
+
+  @NotNull
+  @Override
+  public ProtocolMatcher getMatcher() {
+    return new ProtocolMatcher(Mode.STRICT, getAnnounce(), null);
+  }
+
+  @NotNull
+  @Override
+  public SafeFuture<Controller> initChannel(P2PAbstractChannel channel, String s) {
+    // TODO timeout handlers
+    final Connection connection = ((io.libp2p.core.Stream) channel).getConn();
+    final NodeId nodeId = new LibP2PNodeId(connection.getSecureSession().getRemoteId());
+    Controller controller = new Controller(nodeId);
+    if (!channel.isInitiator()) {
+      controller.setRequestHandler(rpcMethod.createIncomingRequestHandler());
+    }
+    channel.getNettyChannel().pipeline().addLast(controller);
+    return controller.activeFuture;
+  }
+
+  static class Controller extends SimpleChannelInboundHandler<ByteBuf> {
+    private final NodeId nodeId;
+    private RpcRequestHandler rpcRequestHandler;
+    private RpcStream rpcStream;
+    private List<ByteBuf> bufferedData = new ArrayList<>();
+
+    protected final SafeFuture<Controller> activeFuture = new SafeFuture<>();
+
+    private Controller(final NodeId nodeId) {
+      this.nodeId = nodeId;
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) {
+      rpcStream = new LibP2PRpcStream(ctx);
+      activeFuture.complete(this);
+    }
+
+    public RpcStream getRpcStream() {
+      return rpcStream;
+    }
+
+    @Override
+    protected void channelRead0(final ChannelHandlerContext ctx, final ByteBuf msg) {
```

Comment:
is the 0 here on the method name on purpose?

Reply:
yeah - its a quirk of the API (https://netty.io/4.0/api/io/netty/channel/SimpleChannelInboundHandler.html):

> Please keep in mind that #channelRead0(ChannelHandlerContext, I) will be renamed to messageReceived(ChannelHandlerContext, I) in 5.0.